### PR TITLE
Add tests for virtual totals

### PR DESCRIPTION
**Pull request by kbrock**
need these tests to address more edge cases

if you notice, the same attribute for the same models are returning different values depending upon what is preloaded.

This set of tests document the current behavior, so future code changes can make the behavior changes more obvious (as they work to getting the same values regardless of the loaded values)

### CODE REVIEW

File: spec/virtual_total_spec.rb [hunk 1]
```diff
@@ -16,16 +16,34 @@
           .to eq([author0, author1, author2].map(&:id))
       end
 
-      it "calculates totals locally" do
-        author0_id = Author.create_with_books(0).id
-        author2_id = Author.create_with_books(2).id
+      it "calculates totals using a query" do
+        author0 = Author.create_with_books(0).reload
+        author2 = Author.create_with_books(2).reload
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Is the `.reload` here just to make sure we are in a fresh state, and then we are only doing the `total_books` query instead of including the `.find` here?

If so, makes sense, just making sure I am understanding the change correctly.

File: spec/virtual_total_spec.rb [hunk 2]
```diff
@@ -245,7 +271,7 @@ class Author < VitualTotalTestBase
       m2 = model_with_children(2)
       mc = m1.class
       expect {
-        ms = mc.select(:id, mc.arel_attribute(:total_books).as("total_books"))
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Huh... why was this here in the first place?

File: spec/virtual_total_spec.rb [hunk 3]
```diff
@@ -328,4 +362,62 @@ def model_with_children(count)
       end
     end
   end
+
+  describe ".virtual_aggregation" do
+    context "with a standard has_many" do
+      let(:authors) { [author, author2, author3]}
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Nitpick:  Space after the `]`

File: spec/virtual_total_spec.rb [hunk 4]
```diff
@@ -35,6 +53,14 @@
           expect(author_query.map(&:total_books)).to match_array([3, 1, 2])
         end.to match_query_limit_of(1)
       end
+
+      it "with no associated records calculates totals with attribute" do
+        Author.create
+        expect do
+          query = Author.select(:id, :total_books).load
+          expect(query.map(&:total_books)).to match_array([0])
+        end.to match_query_limit_of(1)
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Took me a minute to grok this test, but I like that it is added.  👍

File: spec/virtual_total_spec.rb [hunk 5]
```diff
@@ -328,4 +362,62 @@ def model_with_children(count)
...
+
+      it "calculates sum from preloaded association" do
+        author.recently_published_books.load
+        author2.recently_published_books.load
+        author3.recently_published_books.load
+
+        expect do
+          expect(author.sum_recently_published_books_rating).to eq(6)
+          expect(author2.sum_recently_published_books_rating).to eq(5)
+          expect(author3.sum_recently_published_books_rating).to be_nil
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Yeah, this is super important that we have this distinction here when comparing to the one above (results in `0` instead).  Will be important when doing refactoring/fixes.  👍

File: spec/virtual_total_spec.rb [hunk 6]
```diff
@@ -328,4 +362,62 @@ def model_with_children(count)
...
+        end.to match_query_limit_of(0)
+      end
+
+      it "calculates sum from attribute" do
+        authors
+        expect do
+          query = Author.select(:id, :sum_recently_published_books_rating).order(:id).load
+          expect(query.map(&:sum_recently_published_books_rating)).to match_array([6, 5, 0])
+        end.to match_query_limit_of(2)
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: What is the second query happening here?  Figured it would be just one.

File: spec/virtual_total_spec.rb [hunk 7]
```diff
@@ -328,4 +362,62 @@ def model_with_children(count)
...
+      end
+
+      it "with no associated records calculates sum from attribute (and preloaded association)" do
+        authors
+        expect do
+          query = Author.includes(:recently_published_books).select(:id, :sum_recently_published_books_rating).load
+          expect(query.map(&:sum_recently_published_books_rating)).to match_array([6, 5, nil])
+        end.to match_query_limit_of(2)
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Same as above.

Comment (NickLaMuro) [reply spec/virtual_total_spec.rb]: For this one, I could see it being the `.includes`, but I guess that doesn't exist in the one above.

File: spec/virtual_total_spec.rb [hunk 8]
```diff
@@ -265,6 +291,14 @@ def model_with_children(count)
       expect(model_with_children(2).v_total_named_books).to eq(2)
     end
 
+    it "falls back to default when virtual association is not written correctly in ruby" do
+      a = model_with_children(0)
+      # note: this is broken return, but some of our associations return nil instead of [] / none
+      expect(a).to receive(:named_books).and_return(nil)
```
Comment (NickLaMuro) [inline spec/virtual_total_spec.rb]: Hmm, this (and the note above it) confuses me, but I think it is because I am trying to do this review really quick (and failing).  I assume it is okay, just leaving a note for myself really.

File: spec/virtual_total_spec.rb (hunk 1 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: I was worried about the creation in `create_with_books` leaving some of the associations pre populated/loaded

File: spec/virtual_total_spec.rb (hunk 2 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: ikr?

a long time ago (over 2 years?) we were not smart enough to support `select(:virtual_attribute)`

File: spec/virtual_total_spec.rb (hunk 4 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: @NickLaMuro do you want the load outside of the `expect`?

File: spec/virtual_total_spec.rb (hunk 6 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: there is a bug in `sum_recently_published_books_rating` that if the value is `nil` then it makes a second query

File: spec/virtual_total_spec.rb (hunk 7 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: and the answer is the ditto of above

File: spec/virtual_total_spec.rb (hunk 8 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: we have a `try` in our code that handles a `nil` coming back from an association (virtual I presume).

FWIW/ I don't like it.
struggling keeping this in here.

File: spec/virtual_total_spec.rb (hunk 7 above)
Comment (kbrock) [reply spec/virtual_total_spec.rb]: ooh - I extracted. yes this was deceiving

File: spec/virtual_total_spec.rb (hunk 4 above)
Comment (NickLaMuro) [reply spec/virtual_total_spec.rb]: This was mostly a "👍 comment" for what it is worth.  Most of my confusion came from where `[0]` was coming from (instead of `[]`), but then I realized it was part of `Author.create` above...

Didn't need to change, but looks like you already did.

File: spec/virtual_total_spec.rb (hunk 1 above)
Comment (NickLaMuro) [reply spec/virtual_total_spec.rb]: I like it, and I liked that you droped the `.find` in the `.expect`.  Just making sure I was reading it right.  👍
